pkg/controllers/hosteannotation/hosted_annotation_controller.go

```diff
@@ -0,0 +1,143 @@
+		klog.Errorf("Failed to create addon install controller, %v", err)
+		return err
+	}
+	return nil
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, addOnNames ...string) *Reconciler {
+
+	return &Reconciler{
+		client:     mgr.GetClient(),
+		scheme:     mgr.GetScheme(),
+		addOnNames: sets.NewString(addOnNames...),
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r *Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("addon-hosted-annotation-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// watch ManagedCluster as the primary resource
+	err = c.Watch(source.Kind(mgr.GetCache(), &clusterv1.ManagedCluster{}), &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// watch ManagedClusterAddOn as additional resource
+	err = c.Watch(source.Kind(mgr.GetCache(), &addonapiv1alpha1.ManagedClusterAddOn{}), handler.EnqueueRequestsFromMapFunc(
+		handler.MapFunc(func(ctx context.Context, a client.Object) []reconcile.Request {
+			addOn, ok := a.(*addonapiv1alpha1.ManagedClusterAddOn)
+			if !ok {
+				klog.Error("invalid ManagedClusterAddOn object")
+				return []reconcile.Request{}
+			}
+
+			if !r.filterAddOn(addOn) {
+				return []reconcile.Request{}
+			}
+
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Name: addOn.Namespace,
+					},
+				},
+			}
+		}),
+	))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	cluster := &clusterv1.ManagedCluster{}
+	err := r.client.Get(ctx, req.NamespacedName, cluster)
+	if errors.IsNotFound(err) {
+		return ctrl.Result{}, nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var errs []error
+	for _, addonName := range r.addOnNames.List() {
+		if err = r.applyAddOn(ctx, addonName, cluster); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return ctrl.Result{}, errorsutil.NewAggregate(errs)
+}
+
+// filterAddOn returns true if the contoller should take care of the installation of the given add-on.
+func (r *Reconciler) filterAddOn(addOn *addonapiv1alpha1.ManagedClusterAddOn) bool {
+	if addOn == nil {
+		return false
+	}
+
+	return r.addOnNames.Has(addOn.Name)
+}
+
+// applyAddOn only update addon annotation it is missing, if user update the addon resource, it should not be reverted
+func (r *Reconciler) applyAddOn(ctx context.Context, addonName string, cluster *clusterv1.ManagedCluster) error {
+	addOn := &addonapiv1alpha1.ManagedClusterAddOn{}
+	err := r.client.Get(ctx, types.NamespacedName{Name: addonName, Namespace: cluster.Name}, addOn)
+
+	if errors.IsNotFound(err) {
+		return nil
+	}
+
+	_, hostingClusterName := addon.HostedClusterInfo(addOn, cluster)
+
+	if len(hostingClusterName) == 0 {
```

Comment:
The add-on is not always deployed in the hosted mode when the klusterlet is in the hosted mode. Instead The add-on is deployed in the hosted mode only when
- the klusterlet is in the hosted mode;
- the ManagedCluster has an annotation `addon.open-cluster-management.io/enable-hosted-mode-addons: "true"`;
- and it supports the hosted mode;

Reply:
this is checked in HostedClusterInfo func